accepted/0000-improve-command-suggestions.md

```diff
@@ -0,0 +1,104 @@
+```
+npm ERR! missing script: tes
+npm ERR!
+npm ERR! Did you mean one of these?
+npm ERR!     npm run test
+npm ERR!     npm exec -- test
+```
+
+Current behavior of `npm foo` where `foo` is an unknown command:
+
+```
+Top hits for "foo"
+————————————————————————————————————————————————————————————————————————————————
+npm help package-json                                                     foo:50
+npm help exec                                                             foo:16
+npm help npx                                                              foo:16
+npm help folders                                                          foo:15
+npm help scripts                                                           foo:7
+npm help config                                                            foo:7
+npm help init                                                              foo:6
+npm help developers                                                        foo:4
+npm help star                                                             foo:3
+npm help package-lock-json                                                 foo:3
+————————————————————————————————————————————————————————————————————————————————
+(run with -l or --long to see more context)t
+```
+
+Proposed behavior of `npm foo` when both a script named `"foo"` and a package named `foo` with a discoverable bin exist:
+```
+npm ERR! unknown command: foo
+npm ERR!
+npm ERR! Did you mean one of these?
+npm ERR!     npm run foo
+npm ERR!     npm exec -- foo
+```
+
+The current behavior of `npm exec` is ambiguous enough that implementing these suggestions in a reasonable way there is near impossible. This feature would depend on a proposal such as [#336](https://github.com/npm/rfcs/pull/336) having been implemented in order to be able to avoid the current fallbacks built in to `npm exec`.
+
+## Rationale and Alternatives
+
+This is an alternative to allowing unknown commands to fall back to running scripts or executing binaries, the advantages of this implementation are that it does not make assumptions about what the user wants, and it does not take action without the user being explicit.
+
+We do, however, maintain the benefits of helping the user determine what command they actually intended to run.
+
+Improving discoverability of the correct commands is a safer approach than creating ambiguous behavior.
+
+## Implementation
+
+I would suggest that we add a blessed property to errors thrown by commands `suggestions`. In the CLI framework, if a command rejects with an error that has this property, we display the suggestions.
+
+As for discovering the suggestions, we already have logic to find similarly named scripts based on levenshtein distance. This would need to be abstracted to a utility that can be consumed in `run-script` as well as `exec` and `help-search`.
```

Comment:
So if we do this, we could have the commands do something like `throw Object.assign(new Error('unknown command'), { suggestions: listOfEveryAvailableCommand })` and let the Levenshtein distance pick the best ones to prompt?